### PR TITLE
[Auth0 CRUD Part 1] Implement Auth0 CRUD endpoints

### LINKED ISSUE #91: Datastore integration with Auth0

### LINKED ISSUE #119: Determine and implement CRUD endpoint actions for Auth0
### Background

As part of adding an Auth0 auth layer to the data store, we will be decorating a set of API endpoints with a security decorator that indicates what type of action the endpoint is (CRUD).

The decorator will use this information, together with JWT, and information passed to the decorated function (UUID/resource, etc.), to determine what the user is permitted to do.

### Files Affected

The security check for each type of action will ultimately be handled in the `Auth0` class in `dss/util/auth/auth0.py`.

This class defines `_create() _read() _update() _delete()` methods. Each API endpoint will indicate which of the four types of operation it is. So a call to `PUT /bundles/<uuid>` would map to a `put()` function in `dss.api.bundles`. That function would be decorated with a security decorator that indicates it is a `create` operation. Then the security check that would be called would be the `_create()` method of the `Auth0` class in `dss.util.auth`.

The security decorators are added to each API endpoint that needs auth in `dss/api/{files,bundles,*}.py`

From the `scripts/swagger_auth.py` script we can get a set of API endpoints that will be protected:

```
default_auth = {"/files/{uuid}": ["put"],
                "/subscriptions": ["get", "put"],
                "/subscriptions/{uuid}": ["get", "delete"],
                "/collections": ["get", "put"],
                "/collections/{uuid}": ["get", "patch", "delete"],
                "/bundles/{uuid}": ["put", "patch", "delete"]
               }
```

### Work to be Done

Implement the security checks as follows:

Create:
- Any member of an organization can upload/create data

Read:
- Resources/data are public by default
- Resources/data can be restricted access by adding them to FLAC table

Update:
- Need read and create access to update a bundle
- Admins can update bundles

Delete:
- Only admins can delete bundles

### Related Issues

Relates to Auth0 FLAC user stories #93

### Definition of Done

These security checks are implemented in the `Auth0` class in `dss/util/auth/auth0.py`.

### PR DESCRIPTION
**Pull request by chmreid**
Includes stub methods for FLAC table lookups. Implements logic described in #119.

Still requires work to implement FLAC table lookup mechanism.

Closes #119 (implement crud endpoints for Auth0) and closes #91 (datastore integration with Auth0)

Part 1: #130 (adds mixins for FLAC and Auth0AuthZ extensions, adds parameter checks)

Part 2: #135 (handles case of admin action)

Part 3: #136 (expands and updates tests, updates and fixes auth API)

### CODE REVIEW

File: dss/util/auth/auth0.py [hunk 1]
```diff
@@ -41,27 +57,39 @@ def security_flow(self, **kwargs):
...
-        # so that's how we can get requested UUID
+        """Auth checks for 'read' API actions"""
+        # Data is public if there is no FLAC table entry.
+        self._assert_authorized_flac(**kwargs)
+        return
 
     def _update(self, **kwargs):
-        # Requires checking ownership of UUID
+        """Auth checks for 'update' API actions"""
+        try:
```
Comment (chmreid) [inline dss/util/auth/auth0.py]: Using this try/except structure to allow all admins feels a bit awkward. Maybe instead, we can check whether a user is an admin in the `_assert_authorized_flac()` method of the FLAC mixin class. Most actions should probably allow admins explicitly anyway.

File: dss/util/auth/authorize.py [hunk 2]
```diff
@@ -107,5 +107,20 @@ def _assert_authorized_email(self, emails):
         return
 
 
-class Authorize(TokenGroupMixin, TokenEmailMixin):
+class AdminStatusMixin(TokenGroupMixin, TokenEmailMixin):
+    @property
+    def admin_emails(self):
```
Comment (amarjandu) [inline dss/util/auth/authorize.py:128]: this might belong in Config.

Comment (chmreid) [reply dss/util/auth/authorize.py:128]: Ah, thanks. I knew there was a better way to do this...

Comment (chmreid) [reply dss/util/auth/authorize.py:128]: Sorry, misunderstood this to mean there was already a method in Config that did this, but it looks like there is not yet. You're right, that would be a better place for this bit of code.

**Review by amarjandu — APPROVED** [review]
just a small comment, other than that I think that this is ready to merge.

**Comment by chmreid** [comment]
Rebasing onto #132

File: dss/util/auth/authorize.py (hunk 2 above)
Comment (chmreid) [reply dss/util/auth/authorize.py:128]: Fixed with #132 (incorporated into this PR via a rebase)

File: dss/util/auth/auth0.py [hunk 3]
```diff
@@ -5,16 +5,55 @@
...
+    """
+    Mixin class for Auth0 Authorize class to access groups information
+    added to the JWT by the Auth0 AuthZ extension. These are the groups
+    used to determine FLAC access.
+
+    (Note: the Auth0 AuthZ extension adds groups, roles, and permissions,
+    but here we just use groups.)
+    """
+    def _get_auth0authz_claim(self):
+        return "https://dev.ucsc-cgp-redwood.org/auth0"
```
Comment (chmreid) [inline dss/util/auth/auth0.py]: Todo: fix this hard-coded URL to use OIDC audience instead

Comment (chmreid) [reply dss/util/auth/auth0.py]: Done in b04d16a

**Comment by chmreid** [comment]
Fixing up tests and getting them passing is requiring additional tests, will open a new PR for changes made after the point where you approved them (commit adding the Auth0AuthZ Groups mixin class, plus typo fixes, 5f7e62b).

File: dss/util/auth/auth0.py (hunk 1 above)
Comment (chmreid) [reply dss/util/auth/auth0.py]: Fixed with #135 (adds a decorator to allow admins to bypass security checks)

**Pull request merged**